### DISCUSSION TITLE
Simple ways to reorient image

### DISCUSSION THREAD

**Discussion by ptsii** [ANTsX/ANTs#1492] [Q&A] [answered]
I have some images created from a different program that only produces the old analyze format (paired .hdr/.img, and lacking the orientation info in the header).  When I read them into ITKSNAP, it displays them flipped superior-inferior and anterior-posterior (also with the voxel dimensions rounded off - but I know how to correct that).  I've tried to flip these by using ITKSNAP's Reorient Image, after which in ITKSNAP they appear to be correctly oriented.  However, when I save these out and read them back in over the original images (read in from dicoms, and for which I know the orientation is correct), ITKSNAP reports that "These images have different orientation. ITKSNAP will ignore the header in the image you are loading". 

Investigating the headers using ANTS PrintHeader, it seems that ITKSNAP simply changes "Direction Cos Mtx." from:
1 0 0
0 1 0
0 0 1

to:
1 0 0
0 -1 0
0 0 -1

But when these images are read in over the original images (which have the Direction Cos Mtx values of the first example above), ITKSNAP complains and ignores the new image header info, thereby displaying them incorrectly flipped A-P and S-I.  

Perhaps naively, it seems to me that it should be simple to flip the orientation around so that they are displayed correctly in ITKSNAP (and therefore in ANTS and ANTsR).  I'm wondering if there is a tool in ANTS (or ANTsR) that would allow me to do this in a simple way. My preference is to literally just edit the header in some way, rather than use registration tools and then apply transforms, since I know these are the same space and that its stupid AnalyzePro image export issues that are causing the issue.  

FYI I'd be happy to avoid using AnalyzePro for these reasons, but unfortunately there are elegant image processing tools that I really want to be able to use that are not in ANTS/ANTsR.

Thanks for any suggestions anyone might have.

**Comment by ptsii** [comment]
I found reflectImage in ANTsR, however this doesn't simply reorder the data so as to create a mirror image of a particular dimension (which is what I need). It takes a binary image and creates a version that is no longer binary - so it is morphing the image in some way (again - not what I'm after)

**Comment by ntustison** [comment, marked as answer]
Probably something like this:

```
        probabilityArray <- as.array( probabilityImage )
        probabilityArrayDimension <- dim( probabilityImage )
        probabilityArrayFlipped <- probabilityArray[probabilityArrayDimension[1]:1,,]
        probabilityImage <- as.antsImage( probabilityArrayFlipped, reference = probabilityImage )
```

**Reply by ptsii** [reply]
Thanks, I'll see if I can apply that to my images

**Reply by ptsii** [reply]
Holy cow, that's beautiful... Does exactly what I needed. THANK YOU!

For possible use by others in the future, FWIW here's my explanation of what it does (for my lab wiki):

1)  change image into a simple matrix of numbers:
`> Image_data_as_array <- as.array( Image )`
2) find out the dimensions of the matrix (which tells you how many slices there are in each dimension), and save that in a new variable:
`> Image_data_as_array_dimensions <- dim( Image_as_array )`
3) use this variable to flip (mirror image) the data in that dimension:
`> Image_data_as_array_xflipped <- Image_data_as_array[Image_data_as_array_dimensions[1]:1,,]`
4) put this array (with the now-flipped data) back into an image format:
`> Image_xflipped <- as.antsImage( Image_data_as_array_xflipped, reference = Image )`
5)  save this out of R (if you want to):
`> antsImageWrite(Image_xflipped,"Image_xflipped.nii.gz")`

Notes:
-Image_data_as_array_dimensions[1] = the first value in that variable, which is the number of slices in the first (x?) dimension
-Image_data_as_array[Image_as_array_dimensions[1]:1,,] = reorder the data of the first dimension (only) such that it now goes from the highest value (which equals the number of dimensions) down to 1 (i.e., backwards)
-If you want to flip (mirror image) the other dimensions, do the same thing for them, e.g. to flip all the dimensions:

`Image_data_as_array[Image_data_as_array_dimensions[1]:1, Image_as_array_dimensions[2]:1, Image_as_array_dimensions[3]:1]`

**Comment by ptsii** [comment]
I also found PermuteFlipImageOrientationAxes in ANTS, and this works for images read into ITKSNAP.  However, the same images plotted in ANTsR are still incorrectly oriented.  Looking at the headers, PermuteFlipImageOrientationAxes just changes the Direction Cos Mtx. without actually reordering the data.
